feat: added CRUD for flatpages; Closes #776

9921622 opened the pull request:
Flatpage crud can be accessed from the admin dropdown:
![image](https://user-images.githubusercontent.com/39788517/171763578-2360c97f-92a5-4fdb-a8dd-ced08691ef40.png)

tylerecouture reviewed (changes requested):
This is a big one, AWESOME!  Be ready for a bit of back a forth.  A few things right off the bat:

- Squash the two commits and force push.  No need clean-up/fix-up commits in a PR, as mentioned in the lesson 2 a couple days ago.
- Can you move the templates into the utilities app, since that's where we've put all the other flatpage stuff?  I'm not sure how that will affect viewing the pages themselves, i.e. default.html? Might need to fiddle, not sure.
- Let's move the "Custom Pages" section in the Admin menu to be between the course setup and the Advanced sections.
- When I hit "Cancel" on the edit view, it doesn't do anything
- All views/templates should have a base template, for example, see every other page on the website, all badges pages have the same title and icon, with the content rendered below it.  See Campaign templates for examples (or whatever model's CRUD you were using as a reference)
- Form!  We'll need a custom form that is more user friendly:
    - no "Sites" field.  This value should be provided a default and not in the form.  There will only be one site object, so just get it and use it (Site.objects.first() or whatever)
    - no "Enable comments" field (I don't even know what that does?  What is that?  Can look at it later if it's interesting)
    - The "Content" field should use the Summernote WYSIWYG widget (see all other text areas with formattable content, like quest comments, or announcements, etc)
    - We'll customize the helptext too, but I'll worry about that later.
   
There will be more, but lets start there.  This is awesome work!

9921622 commented:
Added the requested changes

tylerecouture commented on src/templates/navbar-fixed-admin-snippet.html:
```diff
@@ -7,9 +7,14 @@
-        <li><a href="https://github.com/bytedeck/bytedeck/blob/develop/CHANGELOG.md"><i class="fa fa-fw fa-history"></i>&nbsp;Changelog</a></li>
\ No newline at end of file
+        <li><a title="Ask a question, report a bug, or request a feature!" href="https://github.com/bytedeck/bytedeck/discussions"><i class="fa fa-fw fa-comments-o"></i>&nbsp;Discussion Forum</a></li>
+        <li><a href="https://github.com/bytedeck/bytedeck/blob/develop/CHANGELOG.md"><i class="fa fa-fw fa-history"></i>&nbsp;Changelog</a></li>
```
Why are you reverting Bytedeck Forum back to Discussion Forum?  Did you gitpull upstream?  Is this branch based from the most recent develop branch? I'm confused what's happening here.

9921622 replied:
I messed up my rebase. So im guessing git pull upstream will default to my messed up rebase

9921622 replied:
fixed

tylerecouture commented on src/utilities/forms.py:
```diff
@@ -87,3 +91,16 @@ def clean(self, data, initial=None):
+
+
+class CustomFlatpageForm(FlatpageForm):
+    content = SummernoteTextField()
```
`content = SummernoteTextField()`
This line isn't necessary I'm pretty sure.  Is this used any where else in this project?  The default field is fine, we just need the widget (unless you've got a specific reason to use this?  Is it from the latest documentation maybe?)

9921622 replied:
I got it from django-summernote github: https://github.com/summernote/django-summernote

in the read-me referencing SummernoteTextField():
![image](https://user-images.githubusercontent.com/39788517/172508969-4a55c890-e52d-4ca8-9ee7-0ca546120ddb.png)

Since I'm not sure if we already have code dealing with it I just added it

tylerecouture replied:
Oh this is good stuff.  Please remove it for now (so it matches our other uses of Summernote in the app) and I added a new issue to look into this, #1012. Thank you!

9921622 replied:
Removed from flatpage form

tylerecouture commented on src/utilities/templates/flatpages/base.html:
```diff
@@ -0,0 +1,14 @@
+{% extends "base_sidebar.html" %}
```
This base template file should match the base template file in Announcements, Badges, and everywhere else, with contextual changes as needed.

9921622 replied:
Fixed: Copied the format of other base.html, only difference was I stripped the block tags of that were never used

tylerecouture commented on src/utilities/templates/flatpages/default.html:
```diff
@@ -1,11 +1,8 @@
-{% extends "base_sidebar.html" %}
```
This will need a few adjustment once you fix you base.html

9921622 replied:
Made a few adjustments to default. 
- Removed unnecessary tags
- removed js code (i think it just adds html tags to text), because summernoteis now being used

tylerecouture commented on src/utilities/templates/flatpages/flatpage-delete.html:
```diff
@@ -0,0 +1,15 @@
+    <form action="" method="post">
+        {% csrf_token %}
+
+        <p>Are you sure you want to delete this custom page?<p>
```
Add some info about the page here, at minimum it's name so they know what they are confirming. Something like: "Are you sure you want to delete the custom flatpage titled: <a href="{{flatpage url here}}> {{flatpage title here}} </a>"

9921622 replied:
Done
![image](https://user-images.githubusercontent.com/39788517/172518031-c0e6853e-b91e-48ad-8a59-fb953b0fdbee.png)

tylerecouture commented on src/utilities/templates/flatpages/flatpage-form.html:
```diff
@@ -0,0 +1,33 @@
+{% block head_title%}
+
+    {% if not update%}
+        Create Page | 
```
This logic should be in the view and passed as a variable, not in the template.  See the Announcements create/update views and form.html as reference.

9921622 replied:
Changed condition to a variable changed by create/update view

tylerecouture commented on src/utilities/templates/flatpages/flatpage-form.html:
```diff
@@ -0,0 +1,33 @@
+            <a href="{% url 'utilities:flatpage-list' %}" role="button" class="btn btn-danger">Cancel</a>
+            <input type="submit" value="Submit" class="btn btn-success"/>
+            {% if update %}
+                <a href="{% url 'utilities:flatpage-delete' pk=flatpage.pk %}" role="button" class="btn btn-danger">Delete</a>
```
Don't have both Cancel and Delete?  The button combo's here don't mathc the rest of the website, take a look at what the other CRUD views do, especially the Announcements (cus it's simple) and copy that.

9921622 replied:
Removed cancel button from form.html

tylerecouture commented on src/utilities/views.py:
```diff
@@ -13,3 +21,58 @@ def videos(request):
+
+
+def edit_post_request(request):
+    post = request.POST.copy()
```
What is this method doing?  There's no docstring, and it seems hanging out here all weirdly, haha

tylerecouture replied:
Ooooh!  I see what you're doing.  There are many ways of doing this, so for consistency, please see if you can override the `get_form_kwargs()` method instead of the `post()` method in the CreateView and UpdateView below.  For reference, see the `CourseAddStudent` method

9921622 replied:
done

tylerecouture reviewed (changes requested):
Looking good!  A few small tweaks requested and should be good to merge.

9921622 commented:
Added the requested changes

tylerecouture commented on src/utilities/forms.py:
```diff
@@ -87,3 +90,20 @@ def clean(self, data, initial=None):
+        }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Can remove overriding of `__init__()`?  It's not doing anything.

9921622 replied:
removed

tylerecouture commented on src/utilities/urls.py:
```diff
@@ -8,4 +9,9 @@
+    path('custom-pages/create/', views.FlatPageCreateView.as_view(), name='flatpage-create'),
+    path('custom-pages/edit/<int:pk>/', views.FlatPageUpdateView.as_view(), name='flatpage-edit'),
+    path('custom-pages/delete/<int:pk>/', views.FlatPageDeleteView.as_view(), name='flatpage-delete'),
 ]
```
For consistency with the rest of the project, please use underscores in page names and view names instead of dashes.

9921622 replied:
tweaked

tylerecouture reviewed (changes requested):
Starting to get nit-picky, so I'll accept the next one I swear!

9921622 commented on src/utilities/templates/flatpages/flatpage-delete.html:
```diff
@@ -0,0 +1,22 @@
+
+{% block head_title%} Confirm Deletion | {% endblock%}
+
+{% block heading_inner %}Delete Custom Page{% endblock %}
```
fixed

9921622 commented on src/utilities/templates/flatpages/flatpage-list.html:
```diff
@@ -0,0 +1,46 @@
+                                <a class="btn btn-warning" href="{% url 'utilities:flatpage-edit' pk=page.pk %}" role="button" title="Edit this page">
+                                    <i class="fa fa-edit"></i>
+                                </a>
+                                <a class="btn btn-danger" href="{% url 'utilities:flatpage-delete' pk=page.pk %}" role="button" title="Delete this page" style="margin-left: 3px">
```
removed style

9921622 commented on src/utilities/forms.py:
```diff
@@ -87,3 +90,20 @@ def clean(self, data, initial=None):
         #         pass
 
 # http://k
```
removed

9921622 commented:
added requested changes

also added a few non reverse match variants of some asserts inside of ViewTestUtilsMixin
- assertRedirectsLoginURL
- assert200URL
- assert404URL

9921622 commented:
Added requested changes
